src/EntityComponentManager.cc

```diff
@@ -78,20 +76,28 @@ class ignition::gazebo::EntityComponentManagerPrivate
   /// \param[in] _entity Entity that has component newly modified
   public: void AddModifiedComponent(const Entity &_entity);
 
-  /// \brief Map of component storage classes. The key is a component
-  /// type id, and the value is a pointer to the component storage.
-  public: std::unordered_map<ComponentTypeId,
-          std::unique_ptr<ComponentStorageBase>> components;
+  /// \brief A class that stores all components and maps entities to their
+  /// component types
+  public: detail::ComponentStorage componentStorage;
```

Comment:
They are pretty similar, and do indeed have some overlap. When I was originally working on this PR, I actually wondered to myself if `entityComponents` is even needed. I think that the answer to this question is no, but for the time being, I decided to leave `entityComponents` as-is since this PR is already quite large.

What if I leave a `TODO` note and/or ticket an issue about this, and then work on replacing `entityComponents` with `componentStorage` in a separate PR after this is merged? That way, we can avoid adding any more changes to this PR.

Reply:
My concern is that with the current implementation, it's not clear to me what's the source of truth when it comes to entities and their components. 

> replacing entityComponents with componentStorage in a separate PR after this is merged? 

It sounds like we may need to keep `entityComponents` and `componentStorage` in sync, and this may not be what's currently happening. This may be resulting in subtle inconsistencies that our tests may not be picking up. Since `componentStorage` is being introduced in this PR, I'd like to understand whether it's needed before merging it in. If it's a replacement for `entityComponents`, then we should straight out remove that. If it serves a different purpose, I'd like to understand how they're related to each other so we can cover our bases. I'm worried that this may be more pervasive than a simple implementation detail that can be fixed on a future PR.

Reply:
`componentStorage` was intended to replace the old component storage, which stored components sequentially in memory: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L81-L84

> My concern is that with the current implementation, it's not clear to me what's the source of truth when it comes to entities and their components.

I think that this is a little confusing because of how the component storage class was changed, so let me try to explain how component storage worked before this PR vs how it works now (or at least, explain them as I understand it :slightly_smiling_face:):

Before this PR, we stored components in the [`ComponentStorageBase`](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.0.0/include/ignition/gazebo/detail/ComponentStorageBase.hh) class, which stored components of a particular type sequentially in memory. A setback with this approach is that for a particular `ComponentStorageBase` instance, we know the type of component being stored, but we don't have any information about which entity owns a particular component instance in the storage. In order to solve this issue, we use `entityComponents` to store the connecting/missing information: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L116-L120

As you can see, for a given entity, we keep track of the component types it has (`ComponentTypeId`) and where to index the memory in the component type's `CompoentStorageBase` instance (`ComponentId`) in order to get the component instance of a particular type that belongs to this entity. Here's an example of how `entityComponents` is used to get a component of a particular type that belongs to an entity: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L606-L623

So, with the previous implementation, we needed `entityComponents` because the component storage alone did not give us enough information when it came to mapping entities to their components.

Now, let's look at the implementation proposed in this PR. The new `ComponentStorage` class maps an entity to all of its components directly, because we are no longer storing component instances sequentially in memory: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh#L128-L145

This also means that we no longer need a separate component storage instance for each component type - now, we only need once component storage instance, since this single instance will manage all components of every type: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/src/EntityComponentManager.cc#L79-L81

Because of this change, we don't need `entityComponents` anymore - if we want to get a component of a particular type that belongs to a specific entity, we don't need another container that connects entities to their component instances. Instead, we can call this method on the new `ComponentStorage` class: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh#L111-L121

So, as you can see, with these new changes, `entityComponents` is essentially a "convenient container" which provides us with a simple way to check what types of components an entity has: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/src/EntityComponentManager.cc#L122-L126

That being said though, I really don't think `entityComponents` is needed anymore, because the original intent/usage of `entityComponents` (connect entities to their component instances in component storage) is now self-contained in the new `ComponentStorage` class.

So, you're probably wondering, "if we don't need `entityComponents` anymore, then why hasn't it been removed?" I haven't removed it yet for a few reasons:
1. I only wanted to change what was absolutely necessary when creating this PR, because I already knew that in order to meet the bare minimum of getting code to compile, a lot of changes would have to be made. So, I figured that changes like these could wait until it came time for review in order to make it easier for the PR reviewers.
2. When I was originally working on this PR, I was _fairly confident_ that we didn't need `entityComponents` anymore, but I wasn't 100% sure. However, now that I have given you this explanation, I am confident that we no longer need `entityComponents` thanks to the new [`ComponentStorage`](https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh) class :grin: 

So, to recap/summarize: in the old implementation, both `ComponentStorageBase` and `entityComponents` were the source of truth - both were needed in order to store components and map which components belonged to which entity. However, with the new implementation, `ComponentStorage` is the source of truth, and `entityComponents` is no longer needed because the new `ComponentStorage` class provides the mapping of entities to their components.

> It sounds like we may need to keep `entityComponents` and `componentStorage` in sync

As a follow-up from my (long) explanation above, if we decide to keep `entityComponents`, then yes, I agree :+1: however, I don't believe that `entityComponents` is needed anymore.

> I'd like to understand how they're related to each other so we can cover our bases

Hopefully the explanation I gave above helps you understand how `ComponentStorage` relates to `EntityComponents` :sweat_smile:

I'll just add one more note regarding "covering our bases": if we do decide to remove `entityComponents`, we'll need to be sure that we still handle updating `entityComponentsDirty` correctly: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L117-L135

This shouldn't be too difficult, but I just wanted to mention it here as a reminder.

Reply:
Thank you for the explanation, very clear! Good to hear we don't need `entityComponents` anymore and can reduce some of the complexity.

Have you started working on removing that on a separate branch or are you waiting to see if the approach here is accepted before working on that?

Reply:
> Have you started working on removing `entityComponents` on a separate branch or are you waiting to see if the approach here is accepted before working on that?

I have not started removing this yet - my plan was to wait for everything else to be reviewed/approved first before making this change, since this PR is already quite large. I can create a new branch to work on this though if you'd like - just let me know!

> The new component storage class offers APIs such as adding / removing entities, creating and fetching components... This is pretty much what the `EntityComponentManager` does. Why do we need another class? Couldn't these maps just be directly inside `EntityComponentManager`?

I guess that we could store these maps in the `EntityComponentManager`, but I'm not a fan of this approach for a few reasons:
1. The `EntityComponentManager` class is already quite large, and adding the functionality from `ComponentStorage` to `EntityComponentManager` directly would make `EntityComponentManager` even more complex. There are also a few lower-level things going on here with a component's `removed` flag, which I think justifies having a new class for the functionality of mapping entities to their components
2. Breaking things down into multiple classes makes it easier for readers of the code to find what they're looking for (and it makes it easier for us as maintainers, too :slightly_smiling_face:)
3. Having a separate class for `ComponentStorage` means we can test this class directly. If the functionality was put directly into `EntityComponentManager`, it would be harder to test it. While we could make the API in `ComponentStorage` part of the public API for `EntityComponentManager` so that it's easier to test, I don't think this would make much sense since the `ComponentStorage` API is "low level" and shouldn't be used by our users directly (they should revert to the "higher level" API provided by `EntityComponentManager`).

> I'm also not clear why the storage class is currently under detail instead of src. It isn't templated or used by any templates.

You're right :slightly_smiling_face: when I was originally working on this class, I had a templated method that was later removed because it wasn't needed. I have moved this over to the regular `include` and `src` directories in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/4a90e86a00445802f2434cda718b1cd17b90ca78

Reply:
marking this as resolved thanks to #943